kep/640-error-pods-rescheduling/README.md

```diff
@@ -0,0 +1,166 @@
+<!-- /toc -->
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+## Summary
+
+This proposal describes a solution to rescheduling pods which encountered some erros.
+
+## Motivation
+
+The current Kubernetes scheduler performs scheduling for a pod only once. Consequently, if a scheduled pod encounters errors during booting or while running, the scheduler won't intervene to reschedule it to other nodes. 
+The in-place restart strategy provided by kubelet only handles local restarts, which often fall short of resolving issues. 
+For instance, a pod might get OOM killed due to insufficient host-node memory, or it could be rejected by kubelet during startup because of mismatches in NUMA or other device requirements. 
+In such cases, local restarts are ineffective. Manual intervention becomes necessary, leading to increased operational complexity.
+
+This situation contradicts Kubernetes' claim of automatic fault tolerance capabilities, diminishing the reliability and high availability of the service. 
+Therefore, we are motivated to develop rescheduling plugins based on the proposed scheduler framework. These plugins can reschedule pods that encounter errors to different nodes, excluding the nodes where they were previously scheduled.
+
+### Goals
+
+* Reschedule already scheduled pods (encounter some errors) to other nodes automatically.
+* Implement pod rescheduling plugins in PreFilter extension point to prefilter nodes that the pod has already scheduled.
+* Pod rescheduling plugin in this proposal has no conflict with default scheduler's plugins.
+
+### Non-Goals
+
+* Reschedule pods controlled by daemonset or other workload resources which don't exist in default k8s controller.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+A pod submitted to the Kubernetes cluster initially operates smoothly, but over time, as the memory on the node hosting the pod becomes insufficient and the pod's memory requirements grow, it eventually gets CrashLoopBackOff due to out-of-memory (OOM) issues. 
+Consequently, we have to manually relocate the pod to other nodes. But this pod may schedule to the same node with the default scheduler.
+
+#### Story 2
+
+After a pod is submitted to the Kubernetes cluster and successfully scheduled, it may encounter startup issues due to mismatches in hardware resources on the node, 
+such as NUMA or device-related information. In such cases, we need to manually reschedule the pod to a different node. But this pod may schedule to the same node with the default scheduler.
+
+## Design Details
+
+Our scheme is to design two components as outlined in the diagram and described below: Rescheduler-controller and Podrescheduling. 
+Users can configure the `scheduling-retries` (maximum rescheduling retry count) in the metadata annotations of workload resources, such as Deployment and StatefulSet. 
+Rescheduler-controller is responsible for monitoring pods with abnormal states in the cluster. For each abnormal pod, 
```

Comment:
using annotations is fine for starters. Do you plan to introduce a new data structure (CRD object) to hold these informations later on?

The approach of recording possible/discarded target nodes is reminiscent of what DRA does: https://github.com/kubernetes/enhancements/pull/3064/files#diff-8fa237d276346416c2aafa209f721e43c9762f59cabec234eafafe01694de3ee

It's worth exploring how this feature can integrate with DRA (maybe leveraging some of the existing data structures?) and/or if it conflicts with it, and, if conflicts, how manage that

Reply:
> Do you plan to introduce a new data structure (CRD object) to hold these informations later on?

Thanks for your suggestion.
Yes, In the next versions, i will develop a CRD (Custom Resource Definition) to replace the current use of annotations.

I store information in the annotations of the associated workload resources (like deployment) for pods, ensuring there are no conflicts with DRA.

Reply:
Elaborating further, this seems quite like a separate controller which kinda adopts and manages pods which aren't owned by a controller (e.g. pod belonging to a daemonset, deployment...). Would it be the case?

If it would be the case, it becomes IMO a hard sell to have this separate controller in parallel to the existing ones, because the obvious question would become "why don't change kubernetes core to make it work with the existing controllers?" which would lead us in the vicinity of https://docs.google.com/document/d/1-wJhiNy84w7tzFdo9HqwTu5DrVSuXFLGTUv8FBiRAAc/edit#heading=h.5mgn2182hlz6